src/index.ts

```diff
@@ -2,152 +2,581 @@ import {
 
+interface IExecutionTimingMetadata {
+  'shell.execute_reply.started': string;
+  'shell.execute_reply': string;
+  execution_failed: string;
+}
 
+// Interfaces
 interface IMode {
   label: string;
   icon: LabIcon;
 }
 
-const ModeIds = ['always', 'never', 'on-error', 'global-timeout', 'custom-timeout'] as const;
-type ModeId = typeof ModeIds[number];
+interface INotifySettings {
+  defaultMode: ModeId;
+  failureMessage: string;
+  mail: boolean;
+  slack: boolean;
+  successMessage: string;
+  defaultThreshold: number | null;
+}
+
+interface ICellMetadata {
+  mode: ModeId;
+  threshold?: string;
+}
+
+interface IInitialResponse {
+  nbmodel_installed: boolean;
+  email_configured: boolean;
+  slack_configured: boolean;
+}
+
+interface ICellNotification {
+  payload: any;
+  timeoutId: number | null;
+  notificationIssued: boolean;
+}
+
+// Constants
+const ModeIds = ['default', 'never', 'on-error', 'custom-timeout'] as const;
+type ModeId = (typeof ModeIds)[number];
 
 const MODES: Record<ModeId, IMode> = {
-  'always': {
-    label: 'Always',
-    icon: bellFilledIcon
-  },
-  'never': {
-    label: 'Never',
-    icon: bellOffIcon
-  },
-  'on-error': {
-    label: 'On error',
-    icon: bellAlertIcon
-  },
-  'global-timeout': {
-    label: 'If longer than global timeout',
-    icon: bellOutlineIcon
+  default: { label: 'Default', icon: bellOutlineIcon },
+  never: { label: 'Never', icon: bellOffIcon },
+  'on-error': { label: 'On error', icon: bellAlertIcon },
+  'custom-timeout': { label: 'Custom Timeout', icon: bellClockIcon },
+};
+
+// Regular expression for validating timeout input
+const TIMEOUT_PATTERN = /^(\d+(\.\d+)?)([smh])$/;
+
+/**
+ * Generates notification data with a custom message
+ */
+const generateNotificationData = (
+  message: string,
+  cell_id: string,
+): Record<string, any> => ({
+  type: 'NOTIFY',
+  payload: {
+    title: message,
+    body: `Cell id: ${cell_id}`,
   },
-  'custom-timeout': {
-    label: 'If longer than %1',
-    icon: bellOutlineIcon // TODO: custom icon with a tiny clock
-  }
-}
+  isProcessed: false,
+  id: `notify-${Math.random().toString(36).substring(2)}`,
+});
 
+/**
+ * Displays configuration warning for unconfigured services
+ */
+const displayConfigWarning = (
+  service: 'Email' | 'Slack',
+  configKey: string,
+  example: string,
+): void => {
+  JupyterNotification.emit(`${service} Not Configured`, 'error', {
+    autoClose: 3000,
+    actions: [
+      {
+        label: 'Help',
+        callback: () =>
+          showErrorMessage(`${service} Not Configured`, {
+            message: `Add a ${service.toLowerCase()} configuration to directory listed under the config section of jupyter --paths (e.g., ~/.jupyter/jupyter_notify_config.json) to enable ${service.toLowerCase()} notifications. Example: \n{\n  "${configKey}": "${example}"}-config"\n}. If you've already configured it, there might be an issue. Please check the terminal for errors and review your setup.`,
```

Comment:
It looks like that part is spurious?

```suggestion
            message: `Add a ${service.toLowerCase()} configuration to directory listed under the config section of jupyter --paths (e.g., ~/.jupyter/jupyter_notify_config.json) to enable ${service.toLowerCase()} notifications. Example: \n{\n  "${configKey}": "${example}"}. If you've already configured it, there might be an issue. Please check the terminal for errors and review your setup.`,
```

![image](https://github.com/user-attachments/assets/4aac4a92-4035-4d1d-a3ed-20a0b2b077f2)

Reply:
The same dialog is displayed when email is not configured and when SMPT server is not running, which for me shows in jupyter-server logs: `SMTP Configuration Error: Failed to instantiate SMTP: [Errno 111] Connection refused`. Can we add another variable in initial response (in addition to `email_configured`) to indicate if the server is available and display another error if it is not?